src/gui/desktopintegration.cpp

```diff
@@ -135,12 +141,36 @@ void DesktopIntegration::setMenu(QMenu *menu)
     if (menu == m_menu)
         return;
 
+#if defined Q_OS_MACOS
+    if (m_menu)
+        delete m_menu;
+
     m_menu = menu;
 
-#ifdef Q_OS_MACOS
     if (m_menu)
         m_menu->setAsDockMenu();
+#elif defined Q_OS_UNIX
+    const bool systemTrayEnabled = m_systrayIcon;
+    if (m_menu)
+    {
+        if (m_systrayIcon)
+        {
```

Comment:
Have you tested this alternative approach? Explicitly deleting the old menu before replacing?

Reply:
> Have you tested this alternative approach? Explicitly deleting the old menu before replacing?

Do you mean doing the following?
```c++
delete m_systrayIcon->contextMenu();
m_systrayIcon->setContextMenu(newMenu);
```

It doesn't seem like a good idea to delete something referred by another thing. But as far as I remember I tried the following and it still didn't do the job correctly:
```c++
auto oldMenu = m_systrayIcon->contextMenu();
m_systrayIcon->setContextMenu(nullptr);
delete old menu;
m_systrayIcon->setContextMenu(newMenu);
```